### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by DeNeutoy**

### CODE REVIEW

File: allennlp/training/metrics.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    A very general abstract class representing a metric which can be
+    accumulated.
+    """
+    def __call__(self, predictions: torch.Tensor,
+                 gold_labels: torch.Tensor,
+                 mask: Optional[torch.Tensor]):
+
+        raise NotImplementedError
+
+    def get_metric(self, reset: bool) -> Dict[str, float]:
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: I think this should return a float, not a dict, and the model should handle the dictionary keys.

Comment (matt-gardner) [inline allennlp/training/metrics.py]: Well, I see that your `F1` metric returns three things, so now I get why you did it this way.  I think it'd be a bit cleaner, though, if this returned `Union[float, Tuple[float, ...]]`, and the mapping to strings still happened in the model.  What if you want to compute several metrics, for instance?  How are you supposed to reconcile naming conflicts in the metrics you're calling?  Much simpler if the model gives them names, instead of the metrics themselves.

File: allennlp/models/semantic_role_labeler.py [hunk 2]
```diff
@@ -42,12 +44,14 @@ class SemanticRoleLabeler(Model):
...
                  stacked_encoder: Seq2SeqEncoder,
-                 initializer: InitializerApplicator) -> None:
+                 initializer: InitializerApplicator,
+                 metric: Optional[Metric] = None) -> None:
         super(SemanticRoleLabeler, self).__init__()
 
         self.vocab = vocab
         self.text_field_embedder = text_field_embedder
         self.num_classes = self.vocab.get_vocab_size("tags")
+        self.f1_metric = metric
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: Shouldn't this just be `self.metric`?

File: allennlp/training/metrics.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/training/metrics.py]: This should also `raise NotImplementedError`

File: allennlp/training/trainer.py [hunk 3]
```diff
@@ -117,10 +125,24 @@ def train(self) -> None:
                     loss = val_output_dict["loss"]
                     val_loss += loss.data.cpu().numpy()
 
-            # TODO(Mark): Add user specified metrics here, maybe a "metrics" key?
-            logger.info("Training Loss: %3f    Validation Loss: %3f ", train_loss, val_loss)
+                val_metrics = self._model.get_metrics(reset=True) or {}
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: You could slightly simplify some of this logic by adding `metrics['Loss'] = train_loss; val_metrics['Loss'] = val_loss`.  Then you can actually make `self._validation_metric` default to `"loss"` like the docstring says it does (assuming you normalize capitalization issues), and not have any special casing for it below.

File: allennlp/training/trainer.py [hunk 4]
```diff
@@ -117,10 +125,24 @@ def train(self) -> None:
...
+            message_template = "Training %s : %3f    Validation %s : %3f "
+            for name, value in metrics.items():
+                logger.info(message_template, name, value, name, val_metrics[name])
+            logger.info(message_template, "Loss", train_loss, "Loss", val_loss)
+
+            if self._validation_metric:
+                this_epoch = val_metrics[self._validation_metric]
+            else:
+                this_epoch = val_loss
+            if len(validation_metric_counts) > self._patience:
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: This `counts` name is confusing to me - maybe `validation_metric_per_epoch`, or `epoch_validation_metrics`, or something?

File: allennlp/training/trainer.py [hunk 5]
```diff
@@ -117,10 +125,24 @@ def train(self) -> None:
...
+                if validation_metric_counts[-self._patience] > this_epoch:
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: This logic isn't correct: [1, 2, 3, 7, 1] will halt with a patience of two, while it should only halt with a patience of one.

File: allennlp/training/trainer.py [hunk 6]
```diff
@@ -145,7 +167,7 @@ def _save_checkpoint(self,
         if is_best:
             logger.info("Best validation performance so far. "
                         "Copying weights to %s/best.th'.", self._serialization_prefix)
-            shutil.copy(model_path, os.path.join(model_path, "best.th"))
+            shutil.copyfile(model_path, os.path.join(self._serialization_prefix, "best.th"))
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Should you also modify `_restore_checkpoint` to load the best model?  Or should that go somewhere else?

File: tests/training/metrics_test.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        f1_measure = F1Measure(null_prediction_label=0)
+        predictions = torch.Tensor([[0.35, 0.25, 0.1, 0.1, 0.2],
+                                    [0.1, 0.6, 0.1, 0.2, 0.0],
+                                    [0.1, 0.6, 0.1, 0.2, 0.0],
+                                    [0.1, 0.5, 0.1, 0.2, 0.0],
+                                    [0.1, 0.2, 0.1, 0.7, 0.0],
+                                    [0.1, 0.6, 0.1, 0.2, 0.0]])
+        # [True Negative, False Negative, True Positive,
+        #  True Positive, True Positive, False Negative]
+        targets = torch.Tensor([0, 5, 1, 0, 3, 0])
```
Comment (matt-gardner) [inline tests/training/metrics_test.py]: I'm confused by this.  How is 5 a valid target index here?

File: tests/training/metrics_test.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    def test_f1_measure_works_for_sequences(self):
+        f1_measure = F1Measure(null_prediction_label=0)
+        predictions = torch.Tensor([[[0.35, 0.25, 0.1, 0.1, 0.2],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0]],
+                                    [[0.35, 0.25, 0.1, 0.1, 0.2],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0]]])
+        # [[True Negative, False Positive, False Positive],
```
Comment (matt-gardner) [inline tests/training/metrics_test.py]: I think multi-class F1 is just a confusing metric that we probably don't need to worry about.  What you really want to do is compute F1 for each (non-null) label independently, and report that, or some aggregation of the individual F1 measures.  We should actually implement a span-based F1 for SRL.  But, all of that can be in a separate PR.  I'd recommend slightly modifying your F1 measure to take the positive label (instead of the null label), and compute F1 of that label.  Then it can be pretty easily be used directly for binary classification, and if a model wants to instantiate multiples of these for each label, and compute some aggregation, they can.

File: tests/training/metrics_test.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import pytest
+import numpy
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.common.checks import ConfigurationError
+from allennlp.training.metrics import CategoricalAccuracy, F1Measure
+
+class MetricsTest(AllenNlpTestCase):
+
+    def test_categorical_accuracy(self):
```
Comment (matt-gardner) [inline tests/training/metrics_test.py]: Might want to have a test here for multiple calls to the metric, and the reset functionality.

File: allennlp/training/trainer.py (hunk 6 above)
Comment (DeNeutoy) [inline allennlp/training/trainer.py]: No - this is for restoring training, which you want to happen from the last saved point. Loading the best model to test on some data is not the use case I had in mind here.

**Comment by DeNeutoy** [comment]
@matt-gardner This is ready for another look. I removed the implementation from the SRL model, as it's going to be considerably more complex now.

File: allennlp/models/model.py [hunk 10]
```diff
@@ -65,6 +69,18 @@ def forward(self, *inputs) -> Dict[str, torch.Tensor]:  # pylint: disable=argume
         """
         raise NotImplementedError
 
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        """
+        Returns a dictionary of metrics. If the ``forward`` call of your model returns a
```
Comment (matt-gardner) [inline allennlp/models/model.py]: I think this docstring is out of date.  It'd be helpful to note here (and/or in class docstring) that the model is expected to call the metrics in `forward`, with the `Metric` objects themselves handling the accumulation across calls to forward.

File: allennlp/training/metrics.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.params import Params
+
+
+class Metric(Registrable):
+    """
+    A very general abstract class representing a metric which can be
+    accumulated.
+    """
+    def __call__(self, predictions: torch.Tensor,
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: `predictions` should be on the next line.

File: allennlp/training/metrics.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+                 gold_labels: torch.Tensor,
+                 mask: Optional[torch.Tensor]):
+
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Remove blank line.

File: allennlp/training/metrics.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        """
+        Parameters
+        ----------
+        predictions : ``torch.Tensor``, required.
+            A tensor of predictions of shape (batch_size, ..., num_classes).
+        gold_labels : ``torch.Tensor``, required.
+            A tensor of integer class label of shape (batch_size, ...). It must be the same
+            shape as the ``predictions`` tensor without the ``num_classes`` dimension.
+        mask: ``torch.Tensor``, optional (default = None).
+                A masking tensor the same size as ``gold_labels``.
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Indentation is off here.

File: allennlp/training/metrics.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        """
+
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Remove blank line.

File: allennlp/training/metrics.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+    @classmethod
+    def from_params(cls, params: Params):
+        metric_type = params.pop_choice("type", cls.list_available())
+        return cls.by_name(metric_type)(**params.as_dict())  # type: ignore
+
+
+@Metric.register("categorical_accuracy")
+class CategoricalAccuracy(Metric):
+    """
+    Categorical Top-K accuracy. Assumes integer labels, with
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: I'd note in here that this is returned as a value from 0-100, not as a value from 0-1.

File: allennlp/training/metrics.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+    @overrides
+    def reset(self):
+        self.correct_count = 0.0
+        self.total_count = 0.0
+
+
+@Metric.register("f1")
+class F1Measure(Metric):
+    """
+    Computes Precision, Recall and F1 with respect to a given ``positive_label`` label.
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: ```s/``positive_label`` label./``positive_label``./```

File: allennlp/training/metrics.py [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+
+class Metric(Registrable):
+    """
+    A very general abstract class representing a metric which can be
+    accumulated.
+    """
+    def __call__(self, predictions: torch.Tensor,
+                 gold_labels: torch.Tensor,
+                 mask: Optional[torch.Tensor]):
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Brief docstring would be nice, where the point is to tell someone who's implementing a `Metric` what they should expect to receive when implementing this method.

File: allennlp/training/metrics.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+        # False Positives: incorrect positively labeled predictions
+        incorrect_non_null_predictions = (argmax_predictions ==
+                                          self._positive_label).float() * negative_label_mask
+        self.false_positives += (incorrect_non_null_predictions * mask).sum()
+
+    def get_metric(self, reset: bool = False):
+        """
+        Returns
+        -------
+        Returns a tuple of the following metrics based on the accumulated count statistics:
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: `s/Returns a/A/`

File: allennlp/training/metrics.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+        precision : float
+        recall : float
+        f1-measure : float
+        """
+        recall = float(self.true_positives) / float(self.true_positives + self.false_negatives + 1e-13)
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: I'm just being OCD here, but when I see this order I _really_ want to move the `recall` line to below the `precision` line, so that the order matches what's in the docstring and the return value.  Feel free to just laugh at me and leave it as it is, though...

File: allennlp/training/trainer.py [hunk 20]
```diff
@@ -1,7 +1,7 @@
 import logging
 import os
 import shutil
-from typing import Optional
+from typing import Optional, List  # pylint: disable=unused-import
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Man, every time I see this it makes me want to switch to python 3.6, so we can just put the variable types inline and remove all of these `disable=unused-import` statements.  Not for this PR, though...

File: allennlp/training/trainer.py [hunk 21]
```diff
@@ -106,6 +113,7 @@ def train(self) -> None:
                 if self._grad_norm:
                     clip_grad_norm(self._model.parameters(), self._grad_norm)
                 self._optimizer.step()
+            metrics = self._model.get_metrics(reset=True) or {"loss": train_loss}
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: I don't think this is right.  You don't want metrics _or_ loss, you want metrics (if any) _and_ loss.  The loss is just another metric, one that's _always_ computed and added to whatever else is already there.

File: allennlp/training/trainer.py [hunk 22]
```diff
@@ -116,11 +124,28 @@ def train(self) -> None:
                     val_output_dict = self._model.forward(**tensor_batch)
                     loss = val_output_dict["loss"]
                     val_loss += loss.data.cpu().numpy()
-
-            # TODO(Mark): Add user specified metrics here, maybe a "metrics" key?
-            logger.info("Training Loss: %3f    Validation Loss: %3f ", train_loss, val_loss)
-            if self._serialization_prefix:
-                self._save_checkpoint(epoch)
+                val_metrics = self._model.get_metrics(reset=True) or {"loss": val_loss}
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Same here.

File: allennlp/training/trainer.py [hunk 23]
```diff
@@ -116,11 +124,28 @@ def train(self) -> None:
...
+
+                message_template = "Training %s : %3f    Validation %s : %3f "
+                for name, value in metrics.items():
+                    logger.info(message_template, name, value, name, val_metrics[name])
+                logger.info(message_template, "Loss", train_loss, "Loss", val_loss)
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: The idea with putting the loss in the metrics dictionary is that you can now just remove this line.

File: allennlp/training/trainer.py [hunk 24]
```diff
@@ -116,11 +124,28 @@ def train(self) -> None:
...
+                        break
+                validation_metric_per_epoch.append(this_epoch)
+                is_best_so_far = this_epoch == max(validation_metric_per_epoch)
+                if self._serialization_prefix:
+                    self._save_checkpoint(epoch, is_best=is_best_so_far)
+            else:
+                message_template = "Training %s : %3f "
+                for name, value in metrics.items():
+                    logger.info(message_template, name, value)
+                logger.info(message_template, "Loss", train_loss)
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Again here, you can remove this line.

File: tests/training/metrics_test.py [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+    def test_f1_measure_works_for_sequences(self):
+        f1_measure = F1Measure(positive_label=0)
+        predictions = torch.Tensor([[[0.35, 0.25, 0.1, 0.1, 0.2],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0]],
+                                    [[0.35, 0.25, 0.1, 0.1, 0.2],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0],
+                                     [0.1, 0.6, 0.1, 0.2, 0.0]]])
+        # [[True Negative, False Positive, False Positive],
+        #  [True Negative, True Positive, False Negative]]
```
Comment (matt-gardner) [inline tests/training/metrics_test.py]: This comment is out of date.

File: allennlp/training/metrics.py [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+
+
+@Metric.register("categorical_accuracy")
+class CategoricalAccuracy(Metric):
+    """
+    Categorical Top-K accuracy. Assumes integer labels, with
+    each item to be classified having a single correct class.
+    """
+    def __init__(self, top_k: int = 1) -> None:
+        self.top_k = top_k
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Unless you really want to commit to these being part of the public API, I'd recommend making them private.  Same with the F1 metric.
